labman/db/tests/test_process.py

```diff
@@ -988,12 +1038,39 @@ def test_create(self):
 
         plates = obs.plates
         self.assertEqual(len(plates), 1)
-        layout = plates[0].layout
-        self.assertEqual(layout[0][0].composition.i5_composition,
-                         PrimerComposition(1523))
-        self.assertEqual(layout[0][0].composition.i7_composition,
-                         PrimerComposition(1524))
-        self.assertIsNone(layout[-1][-1])
+
+        # The big list comprehension below is not generating a layout, just
+        # reading the layout generated by LibraryPrepShotgunProcess.create into
+        # a convenience format.
+        # When LibraryPrepShotgunProcess.create makes obs, it fills the
+        # obs.plates[0].layout property with a list of lists of lists of Wells.
+        # This makes it very hard to create a test case: you have to know the
+        # database id of each of the wells to instantiate all the expected Well
+        # objects, and it is prohibitively time-consuming for the code to
+        # instantiate them in the test case and compare them to all the Well
+        # objects inobs.plates[0].layout.
+        # Because of this, I chose to set up the part of the test that checks
+        # whether the correct i5 and i7 primer has been assigned to the correct
+        # well using a known-good as a list of lists of lists of strings: the
+        # human-readable well id--A1, etc, the i7 primer name, and the
+        # i5 primer name. The known-good, in this format, is stored in
+        # SHOTGUN_PRIMER_LAYOUT. The list comprehension below is simply
+        # extracting those strings from obs.plates[0].layout (into a
+        # convenience variable obs_primer_layout) so that they can be compared
+        # with the known-good via assertListEqual.
+        # In the cases where a given Well object in obs.plates[0].layout is
+        # None, of course you can't access its various nested string
+        # properties, so None is returned instead of the strings.
+        obs_primer_layout = [[[well.well_id if well is not None else None,
```

Comment:
What about rewriting the comprehension to  following? I think it's translated correctly. I've always found nested comprehensions to be hard to grok. But, this is just personal preference, so please do not consider this a blocking comment.

```python
obs_primer_layout = []
for row in obs.plates[0].layout:
    row_detail = []
    for well in row:
        well_detail = []
        well_detail.append(well.well_id if well is not None else None)
        well_detail.append(well.composition.\
                               i7_composition.\
                               primer_set_composition.\
                               external_id)
        well_detail.append(if well is not None else None)
        well_detail.append(well.composition.\
                           i5_composition.\
                           primer_set_composition.\
                           external_id)
        well_detail.append(if well is not None else None)
        row_detail.append(well_detail)
    obs_primer_layout.append(row_detail)```

Reply:
That's a good point--it is only for historical reasons that I did this with a list comprehension, and it doesn't have to stay as one.  I've rewritten it as loops, as you suggested, and I hope it is much clearer; I've left the explanatory comment, though, just in case :)